qabelbox/src/main/java/de/qabel/desktop/repository/sqlite/SqliteContactRepository.kt

```diff
@@ -0,0 +1,344 @@
+                    val lowerWildSearchString = searchString!!.toLowerCase() + "%"
+                    statement.setString(paramIndex++, lowerWildSearchString)
+                    statement.setString(paramIndex++, lowerWildSearchString)
+                    statement.setString(paramIndex, lowerWildSearchString)
+                }
+                statement.executeQuery().use({ resultSet -> return hydrator.hydrateAll(resultSet) })
+            }
+        } catch (e: SQLException) {
+            throw PersistenceException("failed to load all contacts", e)
+        }
+
+    }
+
+    @Throws(PersistenceException::class)
+    private fun findContactIdentityKeys(contactIds: List<Int>): Map<Int, List<String>> {
+        try {
+            val contactIdentityMap = HashMap<Int, MutableList<String>>()
+            database.prepare(
+                    "SELECT c.id, c2.publicKey " +
+                            "FROM " + ContactDB.TABLE_NAME + " " + ContactDB.TABLE_ALIAS + " " +
+                            "JOIN identity_contacts ic ON (c.id = ic.contact_id) " +
+                            "JOIN identity i ON (i.id = ic.identity_id) " +
+                            "JOIN contact c2 ON (c2.id = i.contact_id) " +
+                            "WHERE c.id IN (" + contactIds.joinToString() + ")").use { statement ->
+                statement.executeQuery().use({ results ->
+                    while (results.next()) {
+                        val id = results.getInt(1)
+                        val identityKey = results.getString(2);
+
+                        if (contactIdentityMap.contains(id)) {
+                            contactIdentityMap[id]!!.add(identityKey);
+                        } else {
+                            contactIdentityMap.put(id, mutableListOf(identityKey))
+                        }
+                    }
+                })
+                return contactIdentityMap
+            }
+        } catch (e: SQLException) {
+            e.printStackTrace()
+            throw PersistenceException("Error loading identities for contacts")
+        }
+
+    }
+
+    @Throws(PersistenceException::class)
+    override fun findWithIdentities(searchString: String?): Collection<Pair<Contact, List<Identity>>> {
+
+        val identities = identityRepository.findAll()
+        val contacts = find(searchString)
+        val contactsIDs = contacts.map { contact -> contact.id }
+
+        val contactsDropUrls = dropUrlRepository.findDropUrls(contactsIDs)
+        val contactIdentityKeys = findContactIdentityKeys(contactsIDs)
+
+        val resultList = ArrayList<Pair<Contact, List<Identity>>>(contacts.size)
+        contacts.map { contact ->
+            contactsDropUrls[contact.id]?.forEach { dropUrl -> contact.addDrop(dropUrl) }
+            resultList.add(Pair(contact, if (contactIdentityKeys.contains(contact.id)) {
+                contactIdentityKeys[contact.id]!!.map { identityKey -> identities.getByKeyIdentifier(identityKey) }
```

Comment:
instead of a contains check and all that duplicated code, just create an extension method:
```
fun Map<Integer, String> mappedIdentities(contactId: Integer) =
 getOrNull(contact.id)?.map { identityKey -> identities.getByKeyIdentifier(identityKey)}
    ?: emptyList()
```

And that part would look like:
`resultList.add(Pair(contact, contactIdentityKeys.mappedIdentities(contact.id)))`